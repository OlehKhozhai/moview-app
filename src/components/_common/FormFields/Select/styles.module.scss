@import 'assets/styles/mixins/colors';

.root {
  display: flex;
  flex-flow: column;
  padding-bottom: 30px;
  position: relative;
  cursor: pointer;
}

.label {
  text-transform: uppercase;
  color: $secondary-color;
}

.input {
  margin-top: 15px;
  padding: 15px;
  border-radius: 4px;
  font-size: 22px;
  border: 1px solid transparent;
  background-color: $primary-light-bg;
  color: $primary-color;
  cursor: pointer;
}

.inputError {
  border-color: $primary-border;
}

.dropdown {
  position: absolute;
  top: 92px;
  right: 0;
  z-index: 5;
  padding: 10px 15px;
  background: $primary-dark-bg;
  left: 0;
  border-radius: 5px;
}

.dropdownItem {
  padding: 10px 0;
  font-size: 22px;
  color: $primary-color;
  text-transform: capitalize;
}

.checkbox {
  margin-right: 10px;
}

.errorMessage {
  bottom: 10px;
}
