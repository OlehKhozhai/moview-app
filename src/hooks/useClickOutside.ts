import { RefObject, useEffect, useState } from 'react';

type Props = {
  ref: RefObject<HTMLElement>;
  handler?: keyof GlobalEventHandlersEventMap;
  closeOnOutsideClick?: boolean;
  callback?: () => void;
};

const useClickOutside = ({
  ref,
  callback,
  handler = 'click',
  closeOnOutsideClick = true,
}: Props) => {
  const [isClickOutside, setClickOutside] = useState(false);

  useEffect(() => {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    const handleClick = (e: any) => {
      const targetIsValid = e.target && document.contains(e.target);
      if (ref && ref.current && targetIsValid) {
        const isClickOnRef = ref.current.contains(e.target);

        if (!isClickOnRef && closeOnOutsideClick) {
          setClickOutside(true);
          if (callback) {
            callback();
          }
        }
      }
    };

    if (ref?.current) {
      document.addEventListener(handler, handleClick);
    }
    return () => {
      document.removeEventListener(handler, handleClick);
    };
  }, [handler, ref, closeOnOutsideClick, callback]);

  return [isClickOutside];
};

export default useClickOutside;
