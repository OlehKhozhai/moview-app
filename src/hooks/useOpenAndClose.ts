import React from 'react';

type Props = { onCloseCallback?: () => void } | undefined;

const useOpenAndClose = (props: Props = {}) => {
  const { onCloseCallback } = props;
  const [isOpen, setIsOpen] = React.useState(false);

  const handleClose = React.useCallback(() => {
    if (onCloseCallback) {
      onCloseCallback();
    }

    setIsOpen(false);
  }, [onCloseCallback]);

  const handleOpen = React.useCallback(() => {
    setIsOpen(true);
  }, []);

  const handleToggle = React.useCallback(
    (e?: React.MouseEvent<HTMLElement>) => {
      if (e) {
        e.stopPropagation();
      }

      setIsOpen((prevState) => {
        if (!prevState && onCloseCallback) {
          onCloseCallback();
        }

        return !prevState;
      });
    },
    [onCloseCallback]
  );

  return {
    isOpen,
    onClose: handleClose,
    onOpen: handleOpen,
    onToggle: handleToggle,
  };
};

export default useOpenAndClose;
