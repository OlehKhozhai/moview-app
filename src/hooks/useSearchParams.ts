import { useEffect, useState } from 'react';
import { useHistory } from 'react-router-dom';

import { DROPDOWN_OPTIONS } from 'config';

const useSearchParams = (): string => {
  const [params, setParams] = useState('');

  const {
    replace: historyReplace,
    location: { search: searchQuery },
  } = useHistory();

  useEffect(() => {
    const searchParams = new URLSearchParams(searchQuery);
    const activeTab = searchParams.get('search');
    const sortBy = searchParams.get('sortBy');

    if (!activeTab) {
      searchParams.set('searchBy', 'genres');
      searchParams.set('search', '');
    }

    if (!sortBy) {
      searchParams.set('sortBy', DROPDOWN_OPTIONS[0]);
    }

    historyReplace({ search: searchParams.toString() });
    setParams(searchParams.toString());
  }, [searchQuery, historyReplace]);

  return params;
};

export default useSearchParams;
