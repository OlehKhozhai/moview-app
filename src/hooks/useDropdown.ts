import React from 'react';

import useOpenAndClose from 'hooks/useOpenAndClose';

type Props =
  | {
      defaultValue?: string;
    }
  | undefined;

const useDropdown = (props: Props = {}) => {
  const { defaultValue } = props;
  const [value, setValue] = React.useState(defaultValue || '');
  const { isOpen, onToggle, onClose } = useOpenAndClose();

  const handleOptionClick = React.useCallback(
    (option: string) => {
      setValue(option);
      onClose();
    },
    [onClose]
  );

  return {
    value,
    onOptionClick: handleOptionClick,
    isDropdownVisible: isOpen,
    onToggleDropdownVisibility: onToggle,
    onCloseDropdown: onClose,
  };
};

export default useDropdown;
