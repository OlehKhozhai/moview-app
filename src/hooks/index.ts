import useOpenAndClose from './useOpenAndClose';
import useDropdown from './useDropdown';
import useClickOutside from './useClickOutside';

export default { useOpenAndClose, useDropdown, useClickOutside };
