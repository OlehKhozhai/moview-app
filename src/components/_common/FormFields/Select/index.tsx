import React, { ChangeEvent } from 'react';
import { useField } from 'formik';
import cn from 'classnames';

import ErrorMessage from 'components/_common/ErrorMessage';
import hooks from 'hooks';
import useClickOutside from 'hooks/useClickOutside';
import styles from './styles.module.scss';

type SelectProps = {
  options: ReadonlyArray<string>;
  label: string;
  name: string;
  placeholder: string;
  className?: string;
};

const Select: React.FC<SelectProps> = ({ label, options, placeholder, ...props }) => {
  const [field, meta, { setValue }] = useField({ ...props });
  const dropdownRef = React.useRef<HTMLDivElement>(null);
  const { isOpen, onClose, onOpen } = hooks.useOpenAndClose();
  useClickOutside({ ref: dropdownRef, callback: isOpen ? onClose : undefined });

  const handleOnChange = (e: ChangeEvent<HTMLInputElement>) => {
    setValue([...field.value, e.target.value]);
  };

  return (
    <div className={styles.root} ref={dropdownRef} onClick={onOpen}>
      <label className={styles.label}>{label}</label>

      <input
        className={cn(styles.input, { [styles.inputError]: meta.touched && meta.error })}
        value={field.value.join(', ')}
        placeholder={placeholder}
        disabled
      />

      {isOpen && (
        <ul className={styles.dropdown}>
          {options.map((option) => (
            <li key={option} className={styles.dropdownItem}>
              <input
                type="checkbox"
                className={styles.checkbox}
                value={option}
                checked={field.value.includes(option)}
                onChange={handleOnChange}
              />
              {option}
            </li>
          ))}
        </ul>
      )}

      {meta.touched && meta.error && (
        <ErrorMessage text={meta.error} className={styles.errorMessage} />
      )}
    </div>
  );
};

export default React.memo(Select);
