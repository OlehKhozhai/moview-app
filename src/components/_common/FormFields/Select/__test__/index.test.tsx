import React from 'react';
import formik from 'formik';
import { create } from 'react-test-renderer';

import hooks from 'hooks';
import Select from '../index';

jest.mock('formik', () => ({
  ...jest.requireActual('formik'),
}));

describe('Select component', () => {
  test('should render Select component by default', () => {
    jest.spyOn(formik, 'useField').mockReturnValueOnce([
      { name: 'name', value: ['one', 'two'], onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: ['one', 'two'], touched: true },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);
    jest.spyOn(hooks, 'useOpenAndClose').mockReturnValueOnce({
      isOpen: false,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      onToggle: jest.fn(),
    });

    const component = create(
      <Select placeholder="placeholder" name="name" label="label" options={['one', 'two']} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Select component with open dropdown', () => {
    jest.spyOn(formik, 'useField').mockReturnValueOnce([
      { name: 'name', value: ['one', 'two'], onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: ['one', 'two'], touched: true },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);
    jest.spyOn(hooks, 'useOpenAndClose').mockReturnValueOnce({
      isOpen: true,
      onOpen: jest.fn(),
      onClose: jest.fn(),
      onToggle: jest.fn(),
    });

    const component = create(
      <Select placeholder="placeholder" name="name" label="label" options={['one', 'two']} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  test('should render Select component with error message', () => {
    jest.spyOn(formik, 'useField').mockReturnValueOnce([
      { name: 'name', value: ['one', 'two'], onBlur: jest.fn(), onChange: jest.fn() },
      { initialTouched: false, value: ['one', 'two'], touched: true, error: 'Error message' },
      { setError: jest.fn(), setTouched: jest.fn(), setValue: jest.fn() },
    ]);

    const component = create(
      <Select placeholder="placeholder" name="name" label="label" options={['one', 'two']} />
    );

    expect(component.toJSON()).toMatchSnapshot();
  });
});
