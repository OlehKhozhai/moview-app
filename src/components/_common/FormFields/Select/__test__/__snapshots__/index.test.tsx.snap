// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select component should render Select component by default 1`] = `
<div
  className="root"
  onClick={[MockFunction]}
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input"
    disabled={true}
    placeholder="placeholder"
    value="one, two"
  />
</div>
`;

exports[`Select component should render Select component with error message 1`] = `
<div
  className="root"
  onClick={[Function]}
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input inputError"
    disabled={true}
    placeholder="placeholder"
    value="one, two"
  />
  <p
    className="root errorMessage"
  >
    Error message
  </p>
</div>
`;

exports[`Select component should render Select component with open dropdown 1`] = `
<div
  className="root"
  onClick={[MockFunction]}
>
  <label
    className="label"
  >
    label
  </label>
  <input
    className="input"
    disabled={true}
    placeholder="placeholder"
    value="one, two"
  />
  <ul
    className="dropdown"
  >
    <li
      className="dropdownItem"
    >
      <input
        checked={true}
        className="checkbox"
        onChange={[Function]}
        type="checkbox"
        value="one"
      />
      one
    </li>
    <li
      className="dropdownItem"
    >
      <input
        checked={true}
        className="checkbox"
        onChange={[Function]}
        type="checkbox"
        value="two"
      />
      two
    </li>
  </ul>
</div>
`;
